File: src/exact/knn_exact_serial.c
#include "../../include/exact/knn_exact_serial.h"

void knn_exact_serial_core(const float* corpus, const float* query, int k, int* indices, float* distances, int corpus_length, int query_length, int d) {
    // Allocate memory for the distance matrix D
    float* D = (float*)malloc(corpus_length * query_length * sizeof(float));
    if (!D) {
        fprintf(stderr, "knn_exact_serial_core: Failed to allocate memory for the distance matrix D\n");
        return;
    }

    // Calculate the distance matrix D (squared Euclidean distances)
    distance_square_matrix(corpus, query, D, corpus_length, query_length, d);

    // Temporary arrays for finding the k nearest neighbors
    float* tmp_distances = (float*)malloc(corpus_length * sizeof(float));
    size_t* tmp_indices  = (size_t*)malloc(k * sizeof(size_t));

    if (!tmp_distances || !tmp_indices) {
        fprintf(stderr, "knn_exact_serial_core: Failed to allocate temporary arrays for k-NN\n");
        free(D);
        return;
    }

    // For each query, find the top-k nearest neighbors using GSL's gsl_sort_smallest
    // Its a QuickSelect implementation, so it does't sort the whole array
    for (int q = 0; q < query_length; q++) {
        // Copy the distances of the current query
        memcpy(tmp_distances, &D[q * corpus_length], corpus_length * sizeof(float));

        // Use GSL to find the indices of the k smallest distances
        gsl_sort_float_smallest_index(tmp_indices, k, tmp_distances, 1, corpus_length);

        // Collect the top-k nearest neighbors (sorted)
        for (int i = 0; i < k; ++i) {
            indices[q * k + i] = (int)tmp_indices[i];
            distances[q * k + i] = (float)sqrt( tmp_distances[tmp_indices[i]] );
        }
    }

    free(D);
    free(tmp_distances);
    free(tmp_indices);
}


void knn_exact_serial(const float* corpus, const float* query, int k, int* indices, float* distances, int corpus_length, int query_length, int d, int num_of_blocks) {
    long    max_chunk_length       =   0;
    int     q_start                =   0;
    int     q_chunk_length         =   0;
    float*  query_chunk            =   NULL;
    // Process query chunks/blocks iteratively
    while (q_start < query_length) {
        // Update usable memory status and evaluate max_chunk_length 
        // based on all the memory allocations needs to be done inside knn_exact_serial_core
        max_chunk_length = (get_usable_memory() - 2*corpus_length*sizeof(float) - k*sizeof(size_t)) / ((corpus_length + 1)*sizeof(float));
        
        // Check if max_chunk_length and num_of_blocks are valid:
        if (max_chunk_length == 0) {
            max_chunk_length = 1;
            fprintf(stderr, "knn_exact_serial: Run out of usable memory (usable memory has a margin, so the program may not fail)");
        }
        else if (num_of_blocks > 0 && query_length / num_of_blocks <= max_chunk_length) {
            max_chunk_length = (long)query_length / num_of_blocks;
            max_chunk_length = (max_chunk_length == 0) ? 1 : max_chunk_length;
        }
        // printf("%ld\n", max_chunk_length);

        q_chunk_length = (q_start + max_chunk_length < query_length) ? max_chunk_length : (query_length - q_start);

        // Allocate and load the query chunk
        query_chunk = (float*)&query[q_start * d];

        // The core function fills the indices and distances, block by block:
        knn_exact_serial_core(corpus, query_chunk, k, (indices + q_start * k), (distances + q_start * k), corpus_length, q_chunk_length, d);

        query_chunk = NULL;
        q_start += max_chunk_length;
    }
}

File: julia/knnAlgorithms/knnExactSerial.jl
using StatsBase      # import Pkg; Pkg.add("StatsBase")
using LinearAlgebra  # import Pkg; Pkg.add("LinearAlgebra")
using HDF5           # import Pkg; Pkg.add("HDF5")


# Estimate the amount of usable memory available on the system (Linux specific).  
# Returns:
#  - size of usable memory in bytes.
function get_usable_memory()
    # This should be replaced with a proper memory check for your environment
    # In Linux, use `Sys.free_memory()` or similar system utilities
    return Sys.free_memory() * 0.7  # Change the scalar to determine the usable memory!
end


# Loads data from an HDF5 file.
# `filename`      Path to the HDF5 file.
# `dataset_name`  Name of the dataset within the HDF5 file to load.
# Returns:
#  - Matrix containing the loaded data.
function load_hdf5(filename::String, dataset_name::String)
    file = h5open(filename, "r")     # Explicitly open the file
    data = read(file[dataset_name])  # Read the dataset
    close(file)                      # Explicitly close the file
    return data
end


# Core function to compute the k-nearest neighbors (k-NN) between two matrices using a serial approach.
# `corpus`: Matrix where each row is a data point from the reference set.
# `query`:  Matrix where each row is a data point from the query set.
# `k`:      The number of nearest neighbors to identify for each query point.
# Returns: 
#  - `indices`   of the k-nearest neighbors for each query (length `query_length x k`)
#  - `distances` to the k-nearest neighbors (*squared*) (length `query_length x k`)
function knn_exact_serial_core(corpus, query, k)
    query_length, _ = size(query)       # Number of rows (samples) in the query

    # Preallocate the result matrices
    indices = Matrix{Int}(undef, query_length, k)
    distances = Matrix{Float32}(undef, query_length, k)

    # Step 1: Compute squared norms for each row in the corpus and query
    corpus_norms = sum(corpus.^2, dims=2)  # Sum of squares for each corpus row
    query_norms = sum(query.^2, dims=2)    # Sum of squares for each query row

    # Step 2: Compute the distance matrix
    dist_matrix = -2 * (query * corpus')
    dist_matrix .+= corpus_norms' .+ query_norms # corpus_norms and query_norms are column vectors


    # Step 3: Find the k nearest neighbors using "partialsortperm"
    for i in 1:query_length
        # Extract the distances for the i-th query
        dists = dist_matrix[i, :]

        # Get the k smallest distances and their indices
        sorted_indices = partialsortperm(dists, 1:k)
        sorted_dists = dists[sorted_indices]
        
        # Store the indices and distances
        indices[i, :] .= sorted_indices
        distances[i, :] .= sorted_dists
    end

    return indices, distances
end


# Parameters:
# - `corpus`            A matrix of data points that form the reference dataset.
# - `query`             A matrix of data points that need neighbors identified.
# - `k`                 The number of nearest neighbors to find for each query point.
# - `num_of_threads`    Number of threads which run this function simultaneously.
function knn_exact_serial(corpus, query, k, num_of_threads)
    println("knn_exact_serial: Start")
    corpus_length, d = size(corpus)
    query_length = size(query, 1)

    # Preallocate result arrays for the entire dataset
    final_indices = Matrix{Int}(undef, query_length, k)
    final_distances = Matrix{Float32}(undef, query_length, k)

    q_start = 1
    while q_start <= query_length
        # Calculate max chunk length
        max_chunk_length = div( (div(get_usable_memory(), num_of_threads) - d * (corpus_length + 1) * sizeof(Float64) - 2 * k * sizeof(Float64)),
                ((corpus_length + d + 2 * k) * sizeof(Float64)) )

        if max_chunk_length == 0
            max_chunk_length = 1
        end
        max_chunk_length = Int(max_chunk_length)
        # println(max_chunk_length)
        
        # Adjust chunk boundaries
        q_chunk_length = min(max_chunk_length, query_length - q_start + 1)

        # Perform k-NN search for the current chunk using core function
        chunk_indices, chunk_distances = knn_exact_serial_core(corpus, query[q_start:q_start + q_chunk_length - 1, :], k)

        # Fill the global indices and distances matrices
        final_indices[q_start:q_start + q_chunk_length - 1, :] .= chunk_indices
        final_distances[q_start:q_start + q_chunk_length - 1, :] .= chunk_distances

        # Move to the next chunk/block
        q_start += max_chunk_length
        println("knn_exact_serial: $(100 * q_start / query_length)% Complete...")
        GC.gc()
    end
    
    println("knn_exact_serial: Done")
    # Squared Distances!
    return final_indices, final_distances
end
